tests/Scheduler/ThreadPool.cpp

```diff
@@ -0,0 +1,350 @@
+  }
+};
+
+template<class T>
+struct PoolBuilder {
+  T makePool(const char* name, unsigned numThreads) {
+    return T{name, numThreads};
+  }
+};
+
+template<>
+struct PoolBuilder<SupervisedSchedulerPool> {
+  tests::mocks::MockRestServer mockApplicationServer;
+
+  SupervisedSchedulerPool makePool(const char*, unsigned numThreads) {
+    return SupervisedSchedulerPool{mockApplicationServer, numThreads};
+  }
+};
+
+template<typename Pool>
+class ThreadPoolPerfTest : public testing::Test {
+  void SetUp() override {
+    arangodb::Logger::CLUSTER.setLogLevel(LogLevel::ERR);
+    arangodb::Logger::THREADS.setLogLevel(LogLevel::ERR);
+  }
+};
+
+using PoolTypes = ::testing::Types<LockfreeThreadPool, WorkStealingThreadPool,
+                                   SupervisedSchedulerPool>;
+TYPED_TEST_SUITE(ThreadPoolPerfTest, PoolTypes);
+
+namespace {
+template<typename Pool>
+struct callable;
+
+template<typename Pool>
+callable<Pool> createLambda(std::atomic<std::uint64_t>& cnt, Pool& pool, int x,
+                            std::atomic<bool>& stop);
+
+template<typename Pool>
+struct callable {
+  void operator()() const noexcept {
+    if (!stop.load()) {
+      cnt.fetch_add(1);
+      pool.push(createLambda(cnt, pool, x + 1, stop));
+      pool.push(createLambda(cnt, pool, x + 1, stop));
+
+      // simulate some work
+      unsigned i = 0;
+      static constexpr unsigned workLimit = 2 << 13;
+      while (++i < workLimit) {
+        // not doing the stop check in every iteration is especially important
+        // when running with TSan!
+        if (i % 1024 == 0 && stop.load()) {
+          break;
+        }
+      }
+    }
+  }
+
+  std::atomic<std::uint64_t>& cnt;
+  std::atomic<bool>& stop;
+  Pool& pool;
+  const int x;
+};
+
+template<typename Pool>
+callable<Pool> createLambda(std::atomic<std::uint64_t>& cnt, Pool& pool, int x,
+                            std::atomic<bool>& stop) {
+  return callable<Pool>{cnt, stop, pool, x};
+}
+}  // namespace
+
+template<typename Pool>
+void spawnWorkTest(unsigned numThreads) {
+  std::atomic<bool> stop = false;
+  std::atomic<std::uint64_t> counter = 0;
+
+  std::uint64_t durationMs = 0;
+  {
+    PoolBuilder<Pool> poolBuilder;
+    auto pool = poolBuilder.makePool("pool", numThreads);
+
+    auto start = std::chrono::steady_clock::now();
+    pool.push(createLambda(counter, pool, 0, stop));
+
+    std::this_thread::sleep_for(std::chrono::seconds{5});
+    stop.store(true);
+    durationMs = std::chrono::duration_cast<std::chrono::milliseconds>(
+                     std::chrono::steady_clock::now() - start)
+                     .count();
+
+    // wait a bit so we don't run into an assertion in the SupervisedScheduler
+    // that we tried to queue an item after the SchedulerFeature was stopped
+    std::this_thread::sleep_for(std::chrono::milliseconds{100});
+  }
+  auto numOps = counter.load();
+  std::cout << std::setw(2) << numThreads << " threads: " << std::setw(5)
```

Comment:
note: this test only prints out timing information, but does not contain any test assertions/expectations.
It is rather a performance test than a functional test. This is fine, but we may want to create separate benchmarks for it rather than running it as part of the functional tests. To be clarified.